client/state/data-layer/http/index.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * External dependencies
+ */
+import superagent from 'superagent';
+
+/***
+ * Internal dependencies
+ */
+import { extendAction } from 'state/utils';
+import { HTTP_REQUEST } from 'state/action-types';
+import { failureMeta, successMeta } from 'state/data-layer/wpcom-http';
+
+const isAllHeadersValid = headers =>
+	headers.every( headerPair => Array.isArray( headerPair ) &&
+			headerPair.length === 2 &&
+			typeof headerPair[ 0 ] === 'string' &&
+			typeof headerPair[ 1 ] === 'string'
+	);
+
+/***
+ * Handler to perform an http request based on `HTTP_REQUEST` action parameters:
+ * {String} url the url to request
+ * {String} method the method we should use in the request: GET, POST etc.
+ * {Array<Array<String>>} headers array of [ 'key', 'value' ] pairs for the request headers
+ * {Object} body data send as body
+ * {Boolean} withCredentials save cookie set on request
+ * {Action} onSuccess action to dispatch on success with data meta
+ * {Action} onFailure action to dispatch on failure with error meta
+ *
+ * @param {Function} dispatch redux store dispatch
+ * @param {Function} action dispatched action we need to handle
+ * @returns {Promise} promise of the handled request
+ */
+const httpHandler = ( { dispatch }, action ) => {
+	const {
+		url,
+		method,
+		body,
+		withCredentials,
+		onSuccess,
+		onFailure
+	} = action;
+
+	const headers = action.headers || [];
+
+	if ( ! isAllHeadersValid( headers ) ) {
+		const error = new Error( "Not all headers were of an array pair: [ 'key', 'value' ]" );
+		dispatch( extendAction( onFailure, failureMeta( error ) ) );
```

Comment:
> typically I would prefer to abort entirely and try to notify the developer immediately that this is wrong. a failed request seems qualitatively different than wrong code.

A user of this code will call it with something like:
```js
export default {
	[ GET_EXAMPLE_DATA ]: [
		dispatchRequest(
			requestExampleData,
			receivedExampleData,
			receivedExampleDataError
	 	)
	],
};
```

if we `throw new Error( '...' )` in the http handler, what will be affected is the `dispatch()` of the `GET_EXAMPLE_DATA`, we'll be throwing at:
```js
return handlerChain.forEach( handler => handler( store, action, localNext ) );
```
at the [middleware](https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom-api-middleware.js#L71)

IMO, that's less useful than hitting the error handler that will generate a proper "error action", which will modify appropriate reducers and be reflected in the UI.

If the headers are wrong in dev time, it'll be obvious immediately and if the bad headers originate in a runtime error we still can tell the user something wrong instead of failing silently with an error in the console.